### PR TITLE
SI-6646 Fix regression in for desugaring.

### PR DESCRIPTION
**Pull request by retronym**
The early check in the parser of pattern irrefutability,
added in c82ecab, failed to consider InitCaps and
`backquoted` identifiers.

Review by @paulp

@adriaanm should take a look to judge the severity.

Given that it's a silent killer, I've formed an opinion that is
expressed with the target branch of this PR.

### CODE REVIEW

**Comment by jsuereth** [comment]
I believe, given the spec and that this is regression, it's a blocker.

**Comment by paulp** [comment]
Looks like you're cleaning up after me all over the place.  LGTM.

**Comment by retronym** [comment]
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/663/console

**Comment by retronym** [comment]
Review once more by @paulp, please.

File: src/reflect/scala/reflect/internal/TreeInfo.scala [hunk 1]
```diff
@@ -245,6 +245,57 @@ abstract class TreeInfo {
     isSelfConstrCall(tree1) || isSuperConstrCall(tree1)
   }
 
+  /**
+   * Does this tree represent an irrefutable pattern match
+   * in the position `for { <tree> <- expr }` based only
+   * on information at the `parser` phase?
+   *
+   * This check is not based on
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/TreeInfo.scala]: this check is not based on... WHAT, THEY'RE ENDING THE SEASON THAT WAY?

File: src/reflect/scala/reflect/internal/TreeInfo.scala [hunk 2]
```diff
@@ -245,6 +245,57 @@ abstract class TreeInfo {
...
+   *   foo @ (bar, (`baz`, quux)) // back quoted ident, not at top level
+   *   foo @ (bar, Quux)          // UpperCase ident, not at top level
+   * }}}
+   *
+   * If the pattern is a simple identifier, it is always
+   * a variable pattern. For example, the following
+   * introduce new bindings:
+   *
+   * {{{
+   * for { X <- xs } yield X
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/TreeInfo.scala:276]: https://groups.google.com/forum/?fromgroups=#!topic/scala-internals/qwa_XOw_7Ks

File: src/reflect/scala/reflect/internal/TreeInfo.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/TreeInfo.scala]: The producers have decided this was too controversial and have rounded out the story.

File: src/reflect/scala/reflect/internal/TreeInfo.scala (hunk 2 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/TreeInfo.scala:276]: I've added the link to the comment. Makes (enough) sense:  `__ <-` is like `__ =`, `case __  =>` is the odd one out.

**Comment by paulp** [comment]
Looks good to me after our exciting little side trip to scala-internals.

**Comment by adriaanm** [comment]
Excellent. I agree this should be fixed now. Great docs.

**Pull request merged**